describe texel formats (for storage texture types)

dneto0 opened the pull request:
I'd like to describe textures in more detail so we can more precisely describe the return type on texture builtin functions, and the sampled component type on sampled texture types.
I'd like to point back to a table in the "Texel formats" section introduced in this PR.

mehmetoguzderin commented on wgsl/index.bs:
```diff
@@ -559,39 +621,39 @@ type.
-  %1 = OpTypeImage type(image_storage_type) 0 0 0 2 image_storage_type ReadOnly
+`texture_ro_2d<texel_format>`
+  %1 = OpTypeImage type(texel_format) 0 0 0 2 texel_format ReadOnly
 
```
Seems like this lacks 2D annotation right after `type(texel_format)` (all others do have)

dneto0 replied:
Fixed in dc7f7538a

mehmetoguzderin commented on wgsl/index.bs:
```diff
@@ -559,39 +621,39 @@ type.
-  %1 = OpTypeImage type(image_storage_type) 0 1 0 2 image_storage_type ReadOnly
+`texture_ro_2d_array<texel_format>`
+  %1 = OpTypeImage type(texel_format) 0 1 0 2 texel_format ReadOnly
 
```
Seems like this lacks 2D annotation right after `type(texel_format)` (all others do have)

dneto0 replied:
Fixed in dc7f7538a

dj2 commented on wgsl/index.bs:
```diff
@@ -520,6 +520,76 @@ type.
+Instead, texture builtin functions are used to access the data.
+
+[SHORTNAME] defines texture types which control which operations are available.
+determine the set of usable builtin functions.
```
Is this a note, or a partial sentence?

dneto0 replied:
partial sentence.  This was rewritten.

dj2 commented on wgsl/index.bs:
```diff
@@ -520,6 +520,76 @@ type.
+<table class='data'>
+  <caption>Channel types</caption>
+  <thead>
+    <tr><td>Channel type<td>Number of stored bits<td>Interpetation of stored bits<td>Shader type<td>Shader value
```
I'd suggest changing this to 1 line per `td` to make it easier to read/work with

dneto0 replied:
I did that for the headings.
But I left the rows as-is.  I find it easier to both edit and scan for differences when they're horizontally organized

dj2 commented on wgsl/index.bs:
```diff
@@ -520,6 +520,76 @@ type.
+  <tr><td>16ui<td>16<td>unsigned integer |v| &isinv; {0,...,65535}<td>u32<td> |v| &div; 65535
+  <tr><td>16i<td>16<td>signed integer |v| &isinv; {-32768,...,32767}<td>i32<td> |v| &div; 32768
+  <tr><td>16f<td>16<td>IEEE 754 16-bit floating point value |v|, with 1 sign bit, 5 exponent bits, 10 mantissa bits<td>f32<td>|v|, converted to f32
+  <tr><td>32ui<td>32<td>32-bit signed integer value |v|<td>u32<td>|v|
```
`ui` should be unsigned integer value no?

dj2 commented on wgsl/index.bs:
```diff
@@ -520,6 +520,76 @@ type.
+  <tr><td>16i<td>16<td>signed integer |v| &isinv; {-32768,...,32767}<td>i32<td> |v| &div; 32768
+  <tr><td>16f<td>16<td>IEEE 754 16-bit floating point value |v|, with 1 sign bit, 5 exponent bits, 10 mantissa bits<td>f32<td>|v|, converted to f32
+  <tr><td>32ui<td>32<td>32-bit signed integer value |v|<td>u32<td>|v|
+  <tr><td>32i<td>32<td>32-bit unsigned integer value |v|<td>i32<td>|v|
```
32 bit signed integer value?

Kangz commented on wgsl/index.bs:
```diff
@@ -520,6 +520,76 @@ type.
+A <dfn noexport>texel</dfn> is a scalar or vector used as the smallest independently addressible element of a [=texture=].
+The word *texel* is short for *texture element*.
+
+A <dfn noexport>texture</dfn> is a grid of texels supporting special operations useful for rendering.
```
nit: multi-dimensional grid (it can be 1, 2, 3 or even 4 if when we support 2D array multisampled textures)

dneto0 replied:
The bullet points below describe dimensionality, and say it's 1,2,3 dimensions. There is also array-size, when the texture is arrayed.

Note that the compute shader workgroup-grid is 3-dimensional without explicitly calling it out as multi-dimensional.

Kangz commented on wgsl/index.bs:
```diff
@@ -520,6 +520,76 @@ type.
+including data layouts, data types, and operations that cannot be expressed directly in the shader language.
+Each texel has a unique grid coordinate within the texture.
+
+An <dfn noexport>arrayed texture</dfn> is a texture which is organized as a homegeneous array of grids of texels.
```
nit: isn't the arrayness just an extra dimension for the multi-dimensional grid?

dneto0 replied:
I was hoping to treat the array index specially. The regular grid coordinates have interpolation applied to them, but the array index does not.  I suggest we add a note to revisit this editorially

Kangz commented on wgsl/index.bs:
```diff
@@ -520,6 +520,76 @@ type.
+
+A texture has the following features:
+
+: <dfn noexport>texel format</dfn>
```
What about mip levels?

dneto0 replied:
Indeed.  I've yet to capture that.

dneto0 replied:
I added mip levels now.  Thanks!

Kangz commented on wgsl/index.bs:
```diff
@@ -520,6 +520,76 @@ type.
+  <tr><td>8uint<td>8<td>unsigned integer |v| &isinv; {0,...,255}<td>u32<td> |v| &div; 255
+  <tr><td>8sint<td>8<td>signed integer |v| &isinv; {-128,...,127}<td>i32<td> |v| &div; 128
+  <tr><td>16uint<td>16<td>unsigned integer |v| &isinv; {0,...,65535}<td>u32<td> |v| &div; 65535
+  <tr><td>16sint<td>16<td>signed integer |v| &isinv; {-32768,...,32767}<td>i32<td> |v| &div; 32768
```
I don't think that the WGSL spec should talk about the exact bit representation of texture, because there's weird formats like rgb9e5 and compressed ones. Only the component type matters, except for storage textures (which can have an exact bit representation described in this spec).

kvark replied:
I think there is still value in describing the formats we *do* know and guarantee, such as `8unorm`. For `rgb9e5` we can figure something out and say it's more platform-dependent.

Kangz replied:
I'm not sure I understand, should the WGSL spec just tell you that when you sample a texture with floating point component type you get an `f32` and what `f32` you get is described by the texture sampling part of the WebGPU specification? This way WGSL wouldn't care if it is `r8unorm` `rg16float` or `bc1-rgba-unorm`?

kvark replied:
well, I'm saying that it's generally useful that, for example, an r8unorm is one of the values between 0/255 and 255/255. This allows authors to adjust their expectations on sampling, or when writing to the color outputs.
Where exactly this is specified (WGSL or WebGPU spec) is not something that I feel strongly about.

dneto0 replied:
The only reason WGSL needs texel-formats is to parameterize storage textures. The sampled and depth textures do not, and I'd like to eliminate those weirdly-shaped texel formats from the WGSL spec entirely.

See the modifications to the texture-ro and texture-wo sections in this MR.

dneto0 replied:
After removing the texel formats that require SPIR-V capability StorageImageExtendedFormats, we're left with:

* r32uint
* r32sint
* r32float
* rgba8unorm
* rgba8snorm
* rgba8uint
* rgba8sint

I can delete all the other ones from the tables, and make this much simpler.

kvark replied:
where did 32-bit RG and RGBA formats go?

dj2 replied:
Looks like `rgba32float` and `rgba32ui`, `rgba32i` are missing from list above but all of the `rg` formats appear to be in `StorageImageExtendedFormats`.  https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#Image_Format

kvark replied:
Interesting! So this is different from what Vulkan requires. I guess the `StorageImageExtendedFormats` on the SPIR-V side becomes an implementation detail, and we don't need to care about it on the API side.
Please just follow the table we have in the spec - https://gpuweb.github.io/gpuweb/#plain-color-formats

dneto0 replied:
I had been reading the required format support for Vulkan. But I have not been following what was worked out already for WebGPU. 
I'll follow the WebGPU table. Thanks!

kvark commented on wgsl/index.bs:
```diff
@@ -520,6 +520,76 @@ type.
+The word *texel* is short for *texture element*.
+
+A <dfn noexport>texture</dfn> is a grid of texels supporting special operations useful for rendering.
+A texture's representation is typically optimized for those operations, with many details hidden from the programmer
```
it's a bit weird to call them "special operations useful for rendering" and then follow with "those operations", which assumes the users know what they are. Would be good to provide examples here

dneto0 replied:
I wanted to defer those to the details of builtin functions.  That's why later in this introductory section we get into the fact there's no direct access to texture storage and that everything is determined by the set of available texture builtins.

Maybe I can think of something that points forward to the builtins without going into details.

Frankly, I'm concerned about having to describe sampling operations. I'd rather not do that in the WGSL spec.  Like, there's a *lot* of content in Vulkan's "14. Image operations". https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#textures

dneto0 commented:
HTML preview is at https://gistpreview.github.io/?4c60af08a8b1cca4501abecc93beae3f

dneto0 commented:
TODO:  capture the concept of mip-levels.

kvark commented on wgsl/index.bs:
```diff
@@ -565,37 +567,168 @@ as the first parameter.
+See [[#texture-ro]] and [[#texture-wo]].
+The last column uses the format-specific [=channel transfer function=] from the channel formats table.
+
+When the texel format does not have a green, blue, or alpha channel, the
```
We can't have that in the language. Is it truly undefined? Could you make an issue/mini-investigation about this?

dneto0 replied:
I filed #1115 for this.

kvark replied:
Let's have an `Issue(gpuweb/1115): xxx` for it here (or whatever the syntax is)

dneto0 replied:
After that discussion evolved, I've made missing green and blue map to 0, and alpha to 1.

kvark commented on wgsl/index.bs:
```diff
@@ -565,37 +567,168 @@ as the first parameter.
+<table class='data'>
+  <caption>Mapping texel formats to SPIR-V</caption>
+  <thead>
+    <tr><td>Texel format<td>SPIR-V Image Format<td>Enabling SPIR-V capability
```
it seems to me that we'd need to remove any mention of this SPIR-V capability, because it doesn't translate to what WebGPU can support, see https://gpuweb.github.io/gpuweb/#texture-format-caps

dneto0 replied:
This table is for translation _to_ SPIR-V.  Some of the image formats that I think are to be supported by WebGPU  do require the extended image formats capability.

kvark replied:
I don't think any of our formats require this capability today.

dneto0 replied:
Ah, ok.  I've updated to eliminate all those extended formats.

kvark commented on wgsl/index.bs:
```diff
@@ -635,42 +768,68 @@ TODO: the RGBA8UNORM-SRGB format is more complicated than this description.
+of the shader value to a stored texel value.
+A write-only storage texture type is parameterized by a [=texel format=], which in turn
+determines the conversion function as specified in [[#texel-formats]].
+The *inverse* of the transfer function is used to convert the shader value to
```
it says "conversion function" and then "transfer function". Is this the same thing?

kvark replied:
it's not clear how the inversion works, i.e. would the values be rounded or truncated, for example

dneto0 commented:
The preview has been updated: https://gistpreview.github.io/?4c60af08a8b1cca4501abecc93beae3f#texel-formats

dneto0 commented:
Rebased against main, and I've updated the gist preview

kvark commented:
My comment may have been lost, so repeating it here. Why did you remove RG32 and RGBA32 formats?

dj2 commented:
The `RG32` formats require `StorageImageExtendedFormats`. Looks like the `RGBA32` ones are missing.

https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#Image_Format

kvark commented:
@dj2 see https://github.com/gpuweb/gpuweb/pull/1112#discussion_r498320940

dneto0 commented:
I've made the list of storage texture formats match what's already listed in the WebGPU table.

That adds the Rg32* formats. Even though those have mandatory in Vulkan, they still require the StorageImageExtendedFormats capability in SPIR-V.  So I added back the "SPIR-V enabling capability" column. 

I've updated the HTML preview: https://gistpreview.github.io/?4c60af08a8b1cca4501abecc93beae3f#texel-formats 

I've added TODOs to move the description about how storage texture accesses work down to the texture builtins section.  The write access needs a more explicitly spelled out description.

dneto0 commented:
I've rebased against main, and collapsed the previous contents into a single commit.

dneto0 commented:
I've updated the HTML preview at https://gistpreview.github.io/?4c60af08a8b1cca4501abecc93beae3f 

Note that other texture type changes are already in flight: #1180, #1182  

I'd like to get these changes approved, modulo those other changes, which are basically orthogonal.

dneto0 commented:
I've rebased on top of recent renamings of storage texture types, and:

- added cross-references to WebGPU concepts
- added mip levels to the list of texture properties.
- fixed wordings.

Please take a look.

grorg commented:
Discussed at [2020-11-17 meeting](https://docs.google.com/document/d/1JMqCbDXyuEvjH8GalT8_bEJ9vfejMo4GMIsAVaXnKvI/).

kvark commented on wgsl/index.bs:
```diff
@@ -671,7 +671,179 @@ Any other context.  Evaluates to *P.Read()*, yielding a value of *P*’s pointee
+Instead, use texture builtin functions as follows:
+
+* Within the shader:
+    * Declare a module-scope variable in the `uniform_constant` storage class,
```
Storage class name needs an update.
In WebGPU spec we'd reference those as something like {{GPUStorageClass/"handle'"}}, which allows `bikeshed` to generate a reference link

dneto0 replied:
I'll fix this to 'handle' now.

The  {{Storage classes/handle}} construct requires a bigger change to the storage class table.

It looks like the syntax that will work here is    [=storage classes/handle=].
I'd like to do the whole storage classes table at once, but that will generate a conflict with this PR.  So I'll do that in a followup.

dneto0 replied:
I filed #1232 for followup

mehmetoguzderin reviewed (approved):
Looks great!
